galter_subjects_utils/contrib/lcsh/converter.py

```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 Northwestern University.
+#
+# galter-subjects-utils is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Generic conversion functionality."""
+
+from .scheme import LCSHScheme
+
+
+class LCSHRDMConverter:
+    """Convert LCSH term into RDM subjects dict."""
+
+    def __init__(self, topics):
+        """Constructor.
+
+        :param topics: LCSH Topics iterable
+        :type topics: iterable[dict]
+        """
+        self.topics = topics
+        self.lcsh = LCSHScheme()
+
+    def __iter__(self):
+        """Iterate over terms.
+
+        DEPRECATED.
+        """
+        yield from self.convert()
+
+    def convert(self):
+        """Iterator over converted entries."""
+        for topic in self.topics:
+            entry = self.extract_entry(topic)
+            if entry:
+                yield entry
+
+    def extract_entry(self, topic):
+        """Extract relevant dict from LCSH skos json-ld dict."""
+        # Find id and label
+        id_suffix = topic["@id"]
+        info_entry = next(
+            (
+                e for e in topic["@graph"]
+                if (
+                    e.get("@id").endswith(id_suffix) and
+                    e.get("@type") == "skos:Concept"
+                )
+            ),
+            None
+        )
+
+        if not info_entry:
+            return {}
+
+        entry = {
+            "id": info_entry["@id"].replace("http://", "https://"),
+            "scheme": self.lcsh.name
+        }
+        pref_label = info_entry["skos:prefLabel"]
+        if isinstance(pref_label, dict):
+            subject = pref_label["@value"]
+        elif isinstance(pref_label, list):
+            subject = pref_label[-1]["@value"]
+        else:
+            raise Exception("invalid skos:prefLabel")
+        entry["subject"] = subject
+
+        return entry
+
+
+def find_main_node(topic):
+    """Find main topic node among topic's skos graph."""
+    id_suffix = topic["@id"]
+    node_main = next(
+        (
+            n for n in topic["@graph"]
+            if n.get("@id", "").endswith(id_suffix)
+        ),
+        None
+    )
+    if not node_main:
+        raise Exception("No main node for {id_suffix}.")
+    return node_main
+
+
+def find_deprecation_node(topic, node_main):
+    """Find deprecation node among topic's skos graph."""
+    for node in topic["@graph"]:
+        is_deprecated = node.get("cs:changeReason") == "deprecated"
```

Comment:
This is funny to me, that LCSH actually use the term "deprecated" for removed terms. I wondered if maybe you picked it, and thought it was interesting choice.

Reply:
Ah, I thought it was a good name on their part since it doesn't always mean just removal: sometimes the term should be replaced by another and sometimes it should be removed. It's also consistent with the idea of marking something that is still around in the wild as not to be used in the future. It reflects that reality nicely. :relaxed: